File: 12_remove_duplicates.py
"""
Task 12: Remove duplicate characters from a string (keep first occurrences).
Reads a line and prints string with duplicates removed.
"""
import sys

def remove_duplicates(s: str) -> str:
    seen = set()
    out = []
    for ch in s:
        if ch not in seen:
            seen.add(ch)
            out.append(ch)
    return ''.join(out)

if __name__ == '__main__':
    s = sys.stdin.readline().rstrip('\n')
    print(remove_duplicates(s))

# Time Complexity: O(n)
# Space Complexity: O(k)

File: 12_remove_duplicates.cpp
#include <bits/stdc++.h>
using namespace std;

/*
Task 12: Remove duplicate characters from a string (keep first occurrences).
Reads a line and prints string with duplicates removed.
*/

int main()
{
    string s;
    if (!getline(cin, s))
        return 0;
    vector<bool> seen(256, false);
    for (char c : s)
    {
        if (!seen[(unsigned char)c])
        {
            cout << c;
            seen[(unsigned char)c] = true;
        }
    }
    cout << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)
